cblox_ros/include/cblox_ros/submap_server_inl.h

```diff
@@ -0,0 +1,802 @@
+  }
+}
+
+template<typename SubmapType>
+void SubmapServer<SubmapType>::integratePointcloud(const Transformation& T_G_C,
+                                                   const Pointcloud& ptcloud_C,
+                                                   const Colors& colors,
+                                                   const bool /*is_freespace_pointcloud*/) {
+  // Note(alexmillane): Freespace pointcloud option left out for now.
+  CHECK_EQ(ptcloud_C.size(), colors.size());
+  tsdf_submap_collection_integrator_ptr_->integratePointCloud(T_G_C, ptcloud_C,
+                                                              colors);
+}
+
+template<typename SubmapType>
+void SubmapServer<SubmapType>::intializeMap(const Transformation& T_G_C) {
+  // Just creates the first submap
+  createNewSubmap(T_G_C, ros::Time::now());
+}
+
+template<typename SubmapType>
+bool SubmapServer<SubmapType>::newSubmapRequired() const {
+  return (num_integrated_frames_current_submap_ >
+          num_integrated_frames_per_submap_);
+}
+
+template<typename SubmapType>
+void SubmapServer<SubmapType>::createNewSubmap(const Transformation& T_G_C,
+                                               const ros::Time& timestamp) {
+  // finishing up the last submap
+  if (!submap_collection_ptr_->empty()) {
+    std::thread finish_submap_thread(&SubmapServer<SubmapType>::finishSubmap,
+                                     this, submap_collection_ptr_->getActiveSubmapID());
+    finish_submap_thread.detach();
+  }
+
+  // Creating the submap
+  const SubmapID submap_id =
+      submap_collection_ptr_->createNewSubmap(T_G_C);
+  // Activating the submap in the frame integrator
+  tsdf_submap_collection_integrator_ptr_->switchToActiveSubmap();
+  // Resetting current submap counters
+  num_integrated_frames_current_submap_ = 0;
+
+  // Updating the active submap mesher
+  active_submap_visualizer_ptr_->switchToActiveSubmap();
+
+  // Publish the baseframes
+  visualizeSubmapBaseframes();
+
+  // Time the start of recording
+  submap_collection_ptr_->getActiveSubmapPtr()->startMappingTime(timestamp.toSec());
+
+  if (verbose_) {
+    ROS_INFO_STREAM("Created a new submap with id: "
+                        << submap_id << ". Total submap number: "
+                        << submap_collection_ptr_->size());
+  }
+}
+
+// Note: Specialized for TsdfSubmap
+template<typename SubmapType>
+void SubmapServer<SubmapType>::finishSubmap(const SubmapID& submap_id) {
+  if (submap_collection_ptr_->exists(submap_id)) {
+    typename SubmapType::Ptr submap_ptr =
+        submap_collection_ptr_->getSubmapPtr(submap_id);
+    // stopping the mapping interval
+    submap_ptr->stopMappingTime(ros::Time::now().toSec());
+    // generating ESDF map
+    submap_ptr->generateEsdf();
+    // publishing the old submap
+    publishSubmap(submap_id);
+    ROS_INFO("[CbloxServer] Finished submap %d", submap_id);
+  }
+}
+template<>
+void SubmapServer<TsdfSubmap>::finishSubmap(const SubmapID& submap_id);
```

Comment:
Rather than a template specialization, I would use a virtual function in the submap subclasses (lol). In addition to being more interpretable, this falls nicely inline with what we've done in voxgraph:

https://github.com/ethz-asl/voxgraph/blob/master/voxgraph/include/voxgraph/frontend/submap_collection/voxgraph_submap.h#L55

(although we should make that function virtual and override ;)